File: plugins/capacitor-network-tools/ios/Sources/NetworkToolsPlugin/SshClient.swift
import Foundation
import NMSSH

struct ExecResult {
    let stdout: String
    let stderr: String
    let exitCode: Int
}

class SshClient {
    private var sessions: [String: NMSSHSession] = [:]
    private let lock = NSLock()

    func connect(host: String, port: Int, username: String, password: String?, privateKey: String?) throws -> String {
        let session = NMSSHSession(host: host, port: port, andUsername: username)
        session.connect()

        guard session.isConnected else {
            throw NSError(domain: "SSH", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to connect to \(host):\(port)"])
        }

        // Authenticate
        if let key = privateKey {
            session.authenticate(byInMemoryPublicKey: nil, privateKey: key, andPassword: password)
        } else if let pass = password {
            session.authenticate(byPassword: pass)
        }

        guard session.isAuthorized else {
            session.disconnect()
            throw NSError(domain: "SSH", code: -2, userInfo: [NSLocalizedDescriptionKey: "Authentication failed for \(username)@\(host)"])
        }

        let sessionId = UUID().uuidString
        lock.lock()
        sessions[sessionId] = session
        lock.unlock()
        return sessionId
    }

    func exec(sessionId: String, command: String, timeout: Int) throws -> ExecResult {
        guard let session = getSession(sessionId) else {
            throw NSError(domain: "SSH", code: -3, userInfo: [NSLocalizedDescriptionKey: "Invalid session ID: \(sessionId)"])
        }

        session.channel.requestTerminal = false
        session.channel.type = NMSSHChannelType.exec

        var error: NSError?
        let response = session.channel.execute(command, error: &error, timeout: NSNumber(value: timeout / 1000))

        if let error = error {
            throw error
        }

        return ExecResult(
            stdout: response ?? "",
            stderr: session.channel.lastResponse ?? "",
            exitCode: Int(session.channel.exitStatus)
        )
    }

    func disconnect(sessionId: String) {
        lock.lock()
        let session = sessions.removeValue(forKey: sessionId)
        lock.unlock()
        session?.disconnect()
    }

    func disconnectAll() {
        lock.lock()
        let allSessions = sessions.values
        sessions.removeAll()
        lock.unlock()
        allSessions.forEach { $0.disconnect() }
    }

    // MARK: - SFTP

    func sftpList(sessionId: String, path: String) throws -> [[String: Any]] {
        guard let session = getSession(sessionId) else {
            throw NSError(domain: "SSH", code: -3, userInfo: [NSLocalizedDescriptionKey: "Invalid session ID"])
        }

        let sftp = NMSFTP(session: session)
        sftp.connect()

        guard sftp.isConnected else {
            throw NSError(domain: "SFTP", code: -1, userInfo: [NSLocalizedDescriptionKey: "SFTP connection failed"])
        }

        defer { sftp.disconnect() }

        guard let files = sftp.contentsOfDirectory(atPath: path) as? [NMSFTPFile] else {
            return []
        }

        return files.compactMap { file -> [String: Any]? in
            guard file.filename != "." && file.filename != ".." else { return nil }
            return [
                "name": file.filename ?? "",
                "path": "\(path)/\(file.filename ?? "")",
                "size": file.fileSize,
                "isDirectory": file.isDirectory,
                "modifiedAt": file.modificationDate?.description ?? "",
                "permissions": file.permissions ?? ""
            ]
        }
    }

    func sftpDownload(sessionId: String, remotePath: String) throws -> String {
        guard let session = getSession(sessionId) else {
            throw NSError(domain: "SSH", code: -3, userInfo: [NSLocalizedDescriptionKey: "Invalid session ID"])
        }

        let sftp = NMSFTP(session: session)
        sftp.connect()

        guard sftp.isConnected else {
            throw NSError(domain: "SFTP", code: -1, userInfo: [NSLocalizedDescriptionKey: "SFTP connection failed"])
        }

        defer { sftp.disconnect() }

        guard let data = sftp.contents(atPath: remotePath) else {
            throw NSError(domain: "SFTP", code: -2, userInfo: [NSLocalizedDescriptionKey: "Failed to download file: \(remotePath)"])
        }

        return data.base64EncodedString()
    }

    func sftpUpload(sessionId: String, remotePath: String, base64Data: String) throws {
        guard let session = getSession(sessionId) else {
            throw NSError(domain: "SSH", code: -3, userInfo: [NSLocalizedDescriptionKey: "Invalid session ID"])
        }

        guard let data = Data(base64Encoded: base64Data) else {
            throw NSError(domain: "SFTP", code: -2, userInfo: [NSLocalizedDescriptionKey: "Invalid base64 data"])
        }

        let sftp = NMSFTP(session: session)
        sftp.connect()

        guard sftp.isConnected else {
            throw NSError(domain: "SFTP", code: -1, userInfo: [NSLocalizedDescriptionKey: "SFTP connection failed"])
        }

        defer { sftp.disconnect() }

        let success = sftp.writeContents(data, toFileAtPath: remotePath)
        if !success {
            throw NSError(domain: "SFTP", code: -3, userInfo: [NSLocalizedDescriptionKey: "Failed to upload file: \(remotePath)"])
        }
    }

    private func getSession(_ sessionId: String) -> NMSSHSession? {
        lock.lock()
        let session = sessions[sessionId]
        lock.unlock()
        return session
    }
}

File: android/src/main/java/com/t6x/plugins/networktools/SshClient.kt
package com.t6x.plugins.networktools

import com.jcraft.jsch.*
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.util.UUID
import java.util.Vector
import java.util.concurrent.ConcurrentHashMap

data class ExecResult(val stdout: String, val stderr: String, val exitCode: Int)

class SshClient(private val registry: SessionRegistry) {

    private val sessions = ConcurrentHashMap<String, Session>()

    fun connect(
        host: String,
        port: Int,
        username: String,
        password: String?,
        privateKey: String?
    ): String {
        val session = createJschSession(host, port, username, password, privateKey)
        val sessionId = UUID.randomUUID().toString()
        sessions[sessionId] = session
        registry.register(
            sessionId, mapOf(
                "host" to host,
                "port" to port,
                "username" to username,
                "password" to password,
                "privateKey" to privateKey
            )
        )
        return sessionId
    }

    fun exec(sessionId: String, command: String, timeout: Long): ExecResult {
        val session = sessions[sessionId] ?: reconnect(sessionId)
        registry.touch(sessionId)

        val channel = session.openChannel("exec") as ChannelExec
        channel.setCommand(command)

        val stdout = ByteArrayOutputStream()
        val stderr = ByteArrayOutputStream()
        channel.outputStream = stdout
        channel.setErrStream(stderr)

        channel.connect(timeout.toInt())

        val startTime = System.currentTimeMillis()
        while (!channel.isClosed) {
            if (System.currentTimeMillis() - startTime > timeout) {
                channel.disconnect()
                throw RuntimeException("Command timed out after ${timeout}ms")
            }
            Thread.sleep(100)
        }

        val exitCode = channel.exitStatus
        channel.disconnect()

        return ExecResult(
            stdout = stdout.toString("UTF-8"),
            stderr = stderr.toString("UTF-8"),
            exitCode = exitCode
        )
    }

    fun disconnect(sessionId: String) {
        val session = sessions.remove(sessionId)
        session?.disconnect()
        registry.evict(sessionId)
    }

    fun disconnectAll() {
        sessions.values.forEach { it.disconnect() }
        sessions.clear()
        registry.evictAll()
    }

    // === SFTP ===

    fun sftpList(sessionId: String, path: String): JSONArray {
        val session = sessions[sessionId] ?: reconnect(sessionId)
        registry.touch(sessionId)

        val channel = session.openChannel("sftp") as ChannelSftp
        channel.connect()

        try {
            @Suppress("UNCHECKED_CAST")
            val entries = channel.ls(path) as Vector<*>

            val files = JSONArray()
            for (item in entries) {
                val entry = item as ChannelSftp.LsEntry
                val name = entry.filename
                if (name == "." || name == "..") continue
                val a = entry.attrs
                val fileObj = JSONObject()
                    .put("name", name)
                    .put("path", "$path/$name")
                    .put("size", a.size)
                    .put("isDirectory", a.isDir)
                    .put("modifiedAt", java.time.Instant.ofEpochSecond(a.mTime.toLong()).toString())
                    .put("permissions", a.permissionsString)
                files.put(fileObj)
            }
            return files
        } finally {
            channel.disconnect()
        }
    }

    fun sftpDownload(sessionId: String, remotePath: String): String {
        val session = sessions[sessionId] ?: reconnect(sessionId)
        registry.touch(sessionId)

        val channel = session.openChannel("sftp") as ChannelSftp
        channel.connect()

        try {
            val output = ByteArrayOutputStream()
            channel.get(remotePath, output)
            return android.util.Base64.encodeToString(output.toByteArray(), android.util.Base64.NO_WRAP)
        } finally {
            channel.disconnect()
        }
    }

    fun sftpUpload(sessionId: String, remotePath: String, base64Data: String) {
        val session = sessions[sessionId] ?: reconnect(sessionId)
        registry.touch(sessionId)

        val channel = session.openChannel("sftp") as ChannelSftp
        channel.connect()

        try {
            val data = android.util.Base64.decode(base64Data, android.util.Base64.DEFAULT)
            channel.put(ByteArrayInputStream(data), remotePath)
        } finally {
            channel.disconnect()
        }
    }

    // === Private ===

    private fun createJschSession(
        host: String,
        port: Int,
        username: String,
        password: String?,
        privateKey: String?
    ): Session {
        val jsch = JSch()

        if (!privateKey.isNullOrBlank()) {
            jsch.addIdentity("key", privateKey.toByteArray(), null, null)
        }

        val session = jsch.getSession(username, host, port)

        if (!password.isNullOrBlank()) {
            session.setPassword(password)
        }

        val config = java.util.Properties()
        config["StrictHostKeyChecking"] = "no"
        session.setConfig(config)

        session.timeout = 15000
        session.setServerAliveInterval(30000)
        session.setServerAliveCountMax(3)
        session.connect()

        return session
    }

    private fun reconnect(sessionId: String): Session {
        val params = registry.getParams(sessionId)
            ?: throw IllegalArgumentException("SSH session expired or unknown: $sessionId")

        val host = params["host"] as? String ?: throw IllegalArgumentException("Stored params missing host")
        val port = (params["port"] as? Int) ?: 22
        val username = params["username"] as? String ?: throw IllegalArgumentException("Stored params missing username")
        val password = params["password"] as? String
        val privateKey = params["privateKey"] as? String

        val session = createJschSession(host, port, username, password, privateKey)
        sessions[sessionId] = session
        return session
    }
}
